rats-rs/src/transport/tls/server.rs

```diff
@@ -0,0 +1,367 @@
+            verify: None,
+            stream: Box::new(GetFdDumpImpl),
+            verify_peer: false,
+        }
+    }
+    pub fn with_verify(mut self, verify: SSL_verify_cb) -> Self {
+        self.verify = verify;
+        self
+    }
+    pub fn with_verify_peer(mut self, verify_peer: bool) -> Self {
+        self.verify_peer = verify_peer;
+        self
+    }
+    pub fn with_tcp_stream(mut self, stream: TcpStream) -> Self {
+        self.stream = Box::new(TcpWrapper(stream));
+        self
+    }
+}
+
+#[maybe_async]
+impl GenericSecureTransPortWrite for Server {
+    async fn send(&mut self, bytes: &[u8]) -> Result<()> {
+        if self.ctx.is_none() || self.ssl_session.is_none() {
+            return Err(Error::kind(ErrorKind::OsslCtxOrSessionUninitialized));
+        }
+        let res = unsafe {
+            SSL_write(
+                self.ssl_session.unwrap(),
+                bytes.as_ptr() as *const libc::c_void,
+                bytes.len() as i32,
+            )
+        };
+        if res < 0 {
+            return Err(Error::kind(ErrorKind::OsslSendFail));
+        }
+        Ok(())
+    }
+
+    async fn shutdown(&mut self) -> Result<()> {
+        if let Some(ssl_session) = self.ssl_session {
+            unsafe {
+                SSL_shutdown(ssl_session);
+                SSL_free(ssl_session);
+            }
+        }
+        if let Some(ctx) = self.ctx {
+            unsafe {
+                SSL_CTX_free(ctx);
+            }
+        }
+        Ok(())
+    }
+}
+
+#[maybe_async]
+impl GenericSecureTransPortRead for Server {
+    async fn receive(&mut self, buf: &mut [u8]) -> Result<usize> {
+        if self.ctx.is_none() || self.ssl_session.is_none() {
+            return Err(Error::kind(ErrorKind::OsslCtxOrSessionUninitialized));
+        }
+        let res = unsafe {
+            SSL_read(
+                self.ssl_session.unwrap(),
+                buf.as_mut_ptr() as *mut libc::c_void,
+                buf.len() as i32,
+            )
+        };
+        if res < 0 {
+            return Err(Error::kind(ErrorKind::OsslReceiveFail));
+        }
+        Ok(res as usize)
+    }
+}
+
+#[maybe_async]
+impl GenericSecureTransPort for Server {
+    async fn negotiate(&mut self) -> Result<()> {
+        let ctx = self
+            .ctx
+            .ok_or(Error::kind(ErrorKind::OsslCtxUninitialize))?;
+        if self.verify_callback.is_some() {
+            let mut mode = SslMode::SSL_VERIFY_NONE;
+            mode |= SslMode::SSL_VERIFY_PEER | SslMode::SSL_VERIFY_FAIL_IF_NO_PEER_CERT;
+            unsafe {
+                SSL_CTX_set_verify(ctx, mode.bits(), self.verify_callback);
+            }
+        }
+        let session = unsafe { SSL_new(ctx) };
+        if session.is_null() {
+            return Err(Error::kind(ErrorKind::OsslNoMem));
+        }
+        unsafe {
+            X509_STORE_set_ex_data(
+                SSL_CTX_get_cert_store(ctx),
+                OPENSSL_EX_DATA_IDX.lock().unwrap().get(),
```

Comment:
Can be replaced with:

```suggestion
                *OPENSSL_EX_DATA_IDX,
```